autoload/ale/c.vim

```diff
@@ -91,3 +92,27 @@ function! ale#c#FindCompileCommands(buffer) abort
 
     return ''
 endfunction
+
+" Given a buffer number, return the relevant command from
+" compile_commands.json. Return 0 if none could be found.
+function! ale#c#GetCompileCommand(buffer) abort
+    let l:build_dir = ale#c#FindCompileCommands(a:buffer)
+    if empty(l:build_dir)
+        return 0
+    endif
+    let l:commands = json_decode(readfile(l:build_dir . '/compile_commands.json'))
```

Comment:
As mentioned in the previous pull request, we should try and replace `readfile()` with a command in the command chain which can be executed to read the file. `cat` will work on Unix machines, and `type` will probably work on Windows. `json_decode` is probably quite a bit slower, but there's no easy way around that for now.

Have a look at the linter code which can execute `make -n` now. What I think we should do is read `compile_commands.json` if the option is on, and fall back on parsing `make -n` if no `compile_commands.json` file can be found. The command callback should return the command to get the contents of `compile_commands.json`, and the other function which parses the output should parse JSON files. It's probably enough to look at the first line of output and see if it looks like the start of some JSON to tell the difference between the two types out output.